### PR TITLE
add support for generating cmake files

### PR DESCRIPTION
**Pull request by hardening**
This new cmake module allows to generate cmake package files. This may ease the
porting for cmake projects that are exporting cmake package informations for other
depending projects. The module uses as much as possible the templates provided by
the cmake installation (and so cmake needs to be installed).

### CODE REVIEW

File: mesonbuild/modules/cmake.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+from .. import build, dependencies, mesonlib, mlog
+from ..interpreterbase import permittedKwargs
+from ..interpreter import ConfigurationDataHolder
+
+
+COMPATIBILITIES = ['AnyNewerVersion', 'SameMajorVersion', 'SameMinorVersion', 'ExactVersion']
+
+# Taken from https://github.com/Kitware/CMake/blob/master/Modules/CMakePackageConfigHelpers.cmake
+PACKAGE_INIT_BASE = '''
+####### Expanded from @PACKAGE_INIT@ by configure_package_config_file() #######
```
Comment (sarum9in) [inline mesonbuild/modules/cmake.py]: It looks a little bit misleading that this text is generated by configure_package_config_file()...

Comment (hardening) [reply mesonbuild/modules/cmake.py]: I just took what was done by the corresponding cmake script. Anyway I can make changes that would mention meson.

**Comment by jpakkane** [comment]
This needs a proper test where a project is built and installed and then CMake is run on the result to ensure that the output works correctly. See for example the pkg-config usage tests in `run_unittests.py`.

**Comment by hardening** [comment]
@jpakkane do you mean something that would test installed_files ?

**Comment by jpakkane** [comment]
Yes. Something that, roughly:

 - builds and installs a library + headers + CMake definitions
 - runs CMake, uses it to find the dependency, builds a test executable that uses the library

Unrelated to this, is this something you personally need or more of a nice-to-have kind of thing?

**Comment by hardening** [comment]
@jpakkane well both, I have a cmake based project that I'd like to migrate to meson (well in fact it's already done), but then there's other projects that depend on it and that also rely on cmake, so I need the cmake package files.

**Comment by nioncode** [comment]
Is this still being considered for merging or should we close it?

**Comment by hardening** [comment]
@nioncode yep I need to rebase, and update a test project

File: mesonbuild/modules/cmake.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+COMPATIBILITIES = ['AnyNewerVersion', 'SameMajorVersion', 'SameMinorVersion', 'ExactVersion']
+
+# Taken from https://github.com/Kitware/CMake/blob/master/Modules/CMakePackageConfigHelpers.cmake
+PACKAGE_INIT_BASE = '''
+####### Expanded from \@PACKAGE_INIT\@ by configure_package_config_file() #######
+####### Any changes to this file will be overwritten by the next CMake run ####
+####### The input file was @inputFileName@ ########
+get_filename_component(PACKAGE_PREFIX_DIR "${CMAKE_CURRENT_LIST_DIR}/@PACKAGE_RELATIVE_PATH@" ABSOLUTE)
+'''
```
Comment (ignatenkobrain) [inline mesonbuild/modules/cmake.py]: __[Flake8]__ <sub>posted by Sider</sub>

[W605] invalid escape sequence '\@'

File: mesonbuild/modules/cmake.py [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+            'CMAKE_SIZEOF_VOID_P': (str(self.detect_voidp_size(state.compilers, state.environment)), '')
+        }
+        mesonlib.do_conf_file(template_file, version_file, conf, 'meson')
+
+        res = build.Data(mesonlib.File(True, state.environment.get_scratch_dir(), version_file), pkgroot)
+        return ModuleReturnValue(res, [res])
+
+    def create_package_file(self, infile, outfile, PACKAGE_RELATIVE_PATH, extra, confdata):
+        package_init = PACKAGE_INIT_BASE.replace('@PACKAGE_RELATIVE_PATH@', PACKAGE_RELATIVE_PATH)\
+                        .replace('@inputFileName@', infile)
```
Comment (ignatenkobrain) [inline mesonbuild/modules/cmake.py]: __[Flake8]__ <sub>posted by Sider</sub>

[E126] continuation line over-indented for hanging indent

**Comment by jpakkane** [comment]
The test fail because of this:

```
Traceback (most recent call last):
  File "run_unittests.py", line 958, in test_markdown_files_in_sitemap
    self.assertIn(f, toc)
```

Specifically because you have added a new `.md` file but have not added it to `sitemap.txt` (this causes it to not show up properly on the web site).

**Comment by hardening** [comment]
@jpakkane any idea how to modify `run_unittests.py` so that the cmake unit test doesn't run if `cmake` is not installed (as when running on azure) ?

**Comment by jpakkane** [comment]
With the annotation `@skipIfNoExecutable('cmake')`.

**Comment by jpakkane** [comment]
The test fails on Cygwin because CMake is not installed. The test should have something like this in it immediately after `project()`:

```meson
cmake_bin = find_program('cmake', required: false)
if not cmake_bin.found()
  error('MESON_SKIP_TEST CMake not installed.')
endif
```

Yes, the way this is done is is different for unit tests and project tests.

**Comment by hardening** [comment]
@jpakkane I applied your suggesttion but it doesn't seem to work.

**Comment by jpakkane** [comment]
The problem seems to be this:
```
        cmakebin = dependencies.ExternalProgram('cmake', silent=False)
        p, stdout, _stderr = mesonlib.Popen_safe(cmakebin.get_command() + ['--system-information'])[0:3]
        if p.returncode != 0:
            mlog.log('cmake not found, returnCode={}'.format(p.returncode))
            return False
```

The error message here is not correct. It's not that CMake was not found, it's that it returned the wrong error code (the  ExternalProgram call will fail if CMake is missing). The error message should instead say that CMake's system information could not be obtained and put the stderr and stdout to the message too so the failure can be debugged.

**Comment by hardening** [comment]
Victory, test finally passed !

**Comment by jpakkane** [comment]
Quick! To the Bat-merge!

**Pull request merged**